fundchannel: add the possibility to fund a channel with specific utxos

darosior opened the pull request:
Fixes #2682 .

This PR adds the possibility to specify a list of outpoints (`[\"txid:vout\", \"txid:vout\"]`) as the second parameter of the `fundchannel` call in addition to `'all'` or an amount in satoshis.

rustyrussell commented on wallet/wallet.c:
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
+			    if (!wallet_update_output_status(
+				    w, &available[j]->txid, available[j]->outnum,
+				    output_state_available, output_state_reserved))
+				    fatal("Unable to reserve output");
```
Um, does this mean I crash the daemon if I try to fund two channels at once from the same UTXO?

darosior replied:
I think so, as in the already used function to select outputs : https://github.com/ElementsProject/lightning/blob/35827111ee837fc0118e0190b9e12539adae81b3/wallet/wallet.c#L315-L318
Anyway we consume the whole utxos if they are directly passed to `fundchannel` so one could not fund 2 channels, or am I missing something ?

ZmnSCPxj replied:
I agree with @darosior, line 388 of the new code is `wallet_get_utxos` which ensures that we can only change update status of UTXOs with `output_state_available`.  Will mark as resolved.

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -39,6 +46,37 @@ struct command_result *param_wtx(struct command *cmd,
+	const char **outpoints = tal_arr(cmd, const char*, 0);
+
+	json_for_each_arr(i, curr, tok)
+		tal_arr_expand(&outpoints, json_strdup(tmpctx, buffer, curr));
```
You want to use `outputs` and the context for json_strdup here: it's a nice convention that a container owns all its members.

rustyrussell replied:
In `listfunds` we use `txid` and `output` fields.  Since we have JSON anyway, I suggest sticking with that, rather than using a boutique encoding for outpoints?

darosior replied:
> In listfunds we use txid and output fields. Since we have JSON anyway, I suggest sticking with that, rather than using a boutique encoding for outpoints?

Does it mean that the user should pass an array of objects ? `[{"txid":"....", "output":0}, ...]`

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -39,6 +46,37 @@ struct command_result *param_wtx(struct command *cmd,
+	if (!tal_count(outpoints))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Could not decode any of the outpoints. The utxos should be"
+			" specified as an array of 'txid:output_index'.");
```
Empty array check should probably be left to later.

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -39,6 +46,37 @@ struct command_result *param_wtx(struct command *cmd,
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"No matching utxo was found from the wallet."
+				"You can get a list of the wallet utxos with"
+				" the `listfunds` RPC call.");
```
Parsing here, before asking wallet_select_specific(), makes this code path simpler too.  If you can't decode txid or outnum, you give a meaningful error here.

Just hand wallet_select_specific an array of txids and an array of outnums: a failure means some outpoint wasn't available.

rustyrussell commented on wallet/wallet.c:
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
+	tal_add_destructor2(utxos, destroy_utxos, w);
+
+	for (i = 0; i < tal_count(outpoints); i++)
+	{
```
style nit: `{` belongs on line above.

rustyrussell commented on wallet/wallet.c:
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
+	char **outpoint;
+	unsigned int output_index;
+	struct bitcoin_txid *txid;
+	struct utxo **available = wallet_get_utxos(ctx, w, output_state_available);
```
available should either be allocated off NULL (which by convention, indicates you're explicitly going to take/free it, as you do below), or just off tmpctx and forget about it.

rustyrussell commented on wallet/wallet.c:
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
+
+			if (bitcoin_txid_eq(&u->txid, txid) && u->outnum == output_index) {
+				    tal_arr_expand(&utxos, u);
+			    if (!wallet_update_output_status(
```
Weird indent?

rustyrussell commented on tests/test_connection.py:
```diff
@@ -753,6 +753,20 @@ def test_funding_toolarge(node_factory, bitcoind):
+
+    # Then fund a channel with them
+    utxos = [utxo["txid"] + ":" + str(utxo["output"]) for utxo in l1.rpc.listfunds()["outputs"]]
+    l1.rpc.fundchannel(l2.info["id"], utxos)
```
OK, test that you can't grab some bogus UTXO too?  And maybe an already-consumed UTXO?

darosior replied:
Ok, how could I get a bogus UTXO ?

ZmnSCPxj replied:
Dump 32 bytes from `/dev/random` for txid plus 2 bytes for outnum?

darosior replied:
I didn't see your comment and finally added a test that makes a node funding a channel with utxos it doesn't own, is it OK ? It takes the same code path.

ZmnSCPxj replied:
It is fine.

darosior commented:
@rustyrussell thank you for the review. I rebased to apply most of the requested changes : 
- Parse outpoints before calling the utxo selection function
- Add some test for failing cases

Added a CHANGELOG brag line :)

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+	const jsmntok_t *curr;
+	struct bitcoin_txid **txids;
+	unsigned int **outnums;
+	char **outpoint = tal_arr(tmpctx, char*, 0);
```
nit: it's weird to me that you allocate this one inline with the declaration, but the others (`txids` and `outnums`) below.

darosior replied:
It seemed weird to me too : I was not aware of the convention @rustyrussell pointed out in his review (to not directly allocate a pointer I will free). Since I dont free this one (`tmpctx`) I thought it was ok to allocate it directly.
I have to admit I'm a little bit confused with which style to use each time I declare a variable.

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+	outnums = tal_arr(cmd, unsigned int*, 0);
+	json_for_each_arr(i, curr, tok) {
+		outpoint = tal_strsplit(cmd, json_strdup(cmd, buffer, curr), ":",
+				STR_NO_EMPTY);
```
nit: you should line up the beginning of this with the end of the `(` above

```
tal_strsplit(cmd ..
             STR_NO_EMPTY
```

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+	json_for_each_arr(i, curr, tok) {
+		outpoint = tal_strsplit(cmd, json_strdup(cmd, buffer, curr), ":",
+				STR_NO_EMPTY);
+		if (tal_count(outpoint) < 3 || strlen(outpoint[0]) != 64)
```
Why are you expecting 'outpoint' to be at least 3 here? Why not check that it's exactly 3? Moving 3 and 64 to constants might be a better way of expressing what you're checking as well.

This would be easier to review if you had included an example of the expected rpc call in the commit notes.

darosior replied:
I test the size of the outpoint array (which always ends with NULL and must have 2elements). I put this condition while testing and should have changed it.

niftynei replied:
Ok cool. Moving to JSON format like rusty recommended above will probably change this too :)

darosior replied:
I don't think, since I need to decode the txid and the outpoint out of the jmn token

darosior replied:
Ok, finally made a json_helper for json token to transaction id :D

darosior replied:
Oh, I just saw that you did the same in your external wallet funding https://github.com/ElementsProject/lightning/pull/2672/commits/d8dcbce4485b91a9b4e6f9037270fedf18b19002

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+			break;
+
+		struct bitcoin_txid *txid = tal(tmpctx, struct bitcoin_txid);
+		unsigned int *outnum = tal(tmpctx, unsigned int);
```
It's weird to see `unsigned int`s allocated, but maybe I'm just new to this.

You probably want to allocate these off of the same thing as the array that you're putting them in, either `cmd` or `outnums`.

darosior replied:
I allocated them off `tmpctx` because they are temo variable, however I'm also new to this and am not sure it's the correct way.

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+		if (!bitcoin_txid_from_hex(outpoint[0], strlen(outpoint[0]), txid)
+				|| opt_set_uintval(outpoint[1], outnum))
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"Could not decode the outpoint with txid \"%s\" and "
```
nit: should be aligned with `(` in previous line.

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+			" of 'txid:output_index'.");
+
+	wtx->utxos = wallet_select_specific(cmd, cmd->ld->wallet, txids, outnums);
+	tal_free(txids);
```
if the `command_fail`s at any point previous to this, do we need to call `tal_free` for each of those branches as well?

darosior replied:
No because tal allocated variables passed to this function are off `cmd`, which is desallocated on `comand_error` and then so they are

niftynei commented on common/wallet_tx.c:
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Could not decode all of the outpoints. The utxos should be"
+			" specified as an array of 'txid:output_index'.");
+	if (tal_count(wtx->utxos) == 0)
```
nit: you can also do `!tal_count(wtx->utxos)`

rustyrussell replied:
... I prefer == 0 here, since it's literally a count.

niftynei commented on common/wallet_tx.c:
```diff
@@ -104,3 +166,34 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+	tx->amount = AMOUNT_SAT(0);
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* version + in count + out count + locktime    index + value + script length */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
```
Technically the `out_len` might be longer than `1` but this is probably fine.

```
(8 + 1 + out_len)
```

darosior replied:
That's why it's a variable

niftynei commented on lightningd/opening_control.c:
```diff
@@ -1059,8 +1059,15 @@ static struct command_result *json_fund_channel(struct command *cmd,
+				BITCOIN_SCRIPTPUBKEY_P2WSH_LEN);
+	} else
+		res = wtx_select_utxos(&fc->wtx, *feerate_per_kw,
+				BITCOIN_SCRIPTPUBKEY_P2WSH_LEN, maxheight);
```
nit: line up with `(`

niftynei commented on common/wallet_tx.c:
```diff
@@ -104,3 +166,34 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* version + in count + out count + locktime    index + value + script length */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
+	for (i = 0; i < tal_count(tx->utxos); ++i) {
```
nit: use postfix, not prefix `++i` -> `i++`

niftynei commented on CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Config: Increased default amount for minimal channel capacity from 1k sat to 10k sat.
 - Build: Non-developer builds are now done with "-Og" optimization.
 - JSON API: `pay` will no longer return failure until it is no longer retrying; previously it could "timeout" but still make the payment.
+- JSON API: `fundchannel` now accepts an utxo array to use to fund the channel as the second parameter (in addition to 'all' or an amount in sats)
```
nit: instead of saying `in addition to`, I'd reword this to `instead of` as you can only provide one of the three ('all', satoshis, or a utxo set)

darosior replied:
Ok. I appreciate that you review my english too, since I'm not a native speaker.

rustyrussell commented:
Agree with all @niftynei feedback.  In particular, I think adding a separate optional 'utxos' param is nicer: if not specified, utxos are selected from current wallet utxos as now.  That does allow us to have a change output from UTXOs if we want to (or 'all' if we don't).

darosior commented:
> Lots of little things, nice work overall!
> 
> I think the biggest question is whether or not we want to have this in addition to the 'amount' field. It seems like it'd be nice to be able to specify a utxo set and a total amount to fund as well, which isn't possible with the current design.
> 
> How you do you see people using this @darosior ?

Thank you for the review.
I see people using it in a few corner cases : a traditional user doing `newaddr`, `connect`, and `fundchannel` won't use it at all. Some users having a lot of tiny inputs wanting more control over them could use it. Actually I think most people who'll use it will consume the whole utxo, that's why I made this `command_result` of `command_result` in case the parameter is an array : I thought making another parameter will just make them copy paste the utxo amount.

rustyrussell commented:
Ideally, we'd allow them to specify any combination, but that's a bit ambitious.  I really think a "utxos" arg is best here.

This will combine well with #2713 too

darosior commented:
Ok, let's do this

rustyrussell commented:
Sorry, you really can't do the API this way.  As @niftynei points out, you may well want to specify both utxos and the amount to spend, so it needs to be a separate 'utxos' argument.

And while txid:outnum is neat on the commandline, it's nasty JSON.  You can always create a plugin which does converts from this syntax :)

darosior commented:
> Sorry, you really can't do the API this way. As @niftynei points out, you may well want to specify both utxos and the amount to spend, so it needs to be a separate 'utxos' argument.

When saying "let's do this" I meant "ok I'll work on the separate 'utxos' argument version (I was not pushing the initial implementation without argument ^^), it a translated sentence from french and has apparently not the same signification in english.

> And while txid:outnum is neat on the commandline, it's nasty JSON. You can always create a plugin which does converts from this syntax :)

I just saw your comment now that I finished to work on the new implementation. I finally kept using JSON as you requested (I added a bitcoin_txid json helper), relying on the fact that a txid would always be 64 hex characters long.

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -39,6 +43,63 @@ struct command_result *param_wtx(struct command *cmd,
+									curr->start + 64, curr->size};
+		/* Pass the `:` separator */
+		const jsmntok_t outnum_tok = {JSMN_STRING, txid_tok.end + 1,
+									curr->end, curr->size};
```
This code can read past the end of malformed input. 

Prefer something like this:

```
static bool split_tok(const char *buffer, const jsmntok_t *t, char split, jsmntok_t *a, jsmntok_t *b)
{
        const char *p = memchr(buffer + t->start, split, buffer + t->end);

        if (!p)
                 return false;

        *a = *b = *t;
        a->end = p - 1 - buffer;
        b->start = p + 1 - buffer;
        return true;
}
```

darosior replied:
That was what I had at first, but I need `const` tokens and didnt want to do multiple cast. Is it preferable to cast ?

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+	/* (version + in count + out count + locktime)  (index + value + script length) */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
+	for (i = 0; i < tal_count(utxos); i++) {
+        if (!*utxos[i]->blockheight || *utxos[i]->blockheight > maxheight) {
```
Weird indent?

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+			return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
+	} else {
+        if (!amount_sat_sub(&tx->change, tx->change, fee_estimate))
```
Weird indent again...

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+		tx->amount = total_amount;
+        if (!amount_sat_sub(&tx->amount, tx->amount, fee_estimate))
+			return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
```
Maybe give a clue here as to what the fee estimate was?

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+	} else {
+        if (!amount_sat_sub(&tx->change, tx->change, fee_estimate))
+            return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
```
Fee estimate printing here too?

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* (version + in count + out count + locktime)  (index + value + script length) */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
+	for (i = 0; i < tal_count(utxos); i++) {
```
mildly prefer C-99-style `(for size_t i = 0;...` here.

rustyrussell commented on common/wallet_tx.c:
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
+								"Cannot afford transaction");
+	}
+	res = check_amount(tx, tx->amount);
+    return res;
```
indent is weird, but just `return check_amount...` instead of a temporary variable?

rustyrussell commented:
OK, minor fixes only.  I'm trying to tag rc1, but there are a couple of other merges pending review, so I'll wait (was tempted to tweak it myself, but that's a bad habit!).

darosior commented:
@rustyrussell I corrected the minor fixes, added the split function (as a helper) and feerate estimate in errors, I also split the first check about parsing the txid and vout in two parts to have more accurate errors.

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/commit/035ea5be213c235c2e17ca8b5da867adf789447a

GitHub ordered the commits weirdly, but I checked locally and they're in the expected order...
